apis/python/test/test_module.py

```diff
@@ -13,12 +13,32 @@ def test_tdbMatrix(tmpdir):
     create_array(p, data)
 
     ctx = vspy.Ctx({})
-    m = vspy.tdbColMajorMatrix_f32(ctx, p, 0, 0, 0, 0, 0, 0)
+    # Read all rows and cols automatically.
+    m = vspy.tdbColMajorMatrix_f32(ctx, p, 0, None, 0, None, 0, 0)
     m.load()
     m_array = np.array(m)
     assert m_array.shape == data.shape
     assert np.array_equal(m_array, data)
 
+    # Read all rows and cols by specifying how many rows and cols there are.
+    m = vspy.tdbColMajorMatrix_f32(ctx, p, 0, data.shape[0], 0, data.shape[1], 0, 0)
+    m.load()
+    m_array = np.array(m)
+    assert m_array.shape == data.shape
+    assert np.array_equal(m_array, data)
+
+    # Read all rows and no cols.
+    m_no_fill = vspy.tdbColMajorMatrix_f32(ctx, p, 0, None, 0, 0, 0, 0)
```

Comment:
With the `std::optional` approach there isn't really a way to specify a length, i.e., a `std::nullopt` plus an offset, rather we can only specify an actual `stop`.  If we don't know the start of the domain, we can't compute the size.  This is also a problem for vectors and is also a problem in previous approach with `0` is a sentinel value.  This doesn't seem to have been a problem since all the domains have been starting at `0` (so far) and so `stop` and `size` happen to coincide.

But we similarly don't know the size if we go in the other half-full direction (even if we start at `0`) so this might not be an issue.  We should definitely *not* define the arguments to mean origin + size.